pkg/core/bundle_list.go

```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
+
+	// return at once. Caller may chose to wait on returned WaitGroup
+	return batchChan, errorChan, &wg2
+}
+
+// watchForInterrupts broadcasts a done signal to several output channels
+func watchForInterrupts(doneChan <-chan struct{}, wg *sync.WaitGroup, outputChans ...chan<- struct{}) {
+	defer func() {
+		wg.Done()
+	}()
+
+	if _, interrupt := <-doneChan; interrupt {
+		for _, outputChan := range outputChans {
+			outputChan <- struct{}{}
+		}
+	}
+}
+
+// watchForErrors listens on 2 independent errors channels and report back to a main error channel.
+//
+// NOTE: this one does not need signaling, since any interruption bubbles up as error.
+func watchForErrors(
+	errorChan1, errorChan2 <-chan error,
+	outputChan chan<- error,
+	doneChan1, doneChan2 chan<- struct{}, wg *sync.WaitGroup) {
+	defer func() {
+		close(outputChan)
+		wg.Done()
+	}()
+
+	var (
+		err                            error
+		errorChan1Open, errorChan2Open bool = true, true
+	)
+	for errorChan1Open || errorChan2Open { // wait on both channels to close
+		select {
+		case err, errorChan1Open = <-errorChan1:
+			if errorChan1Open {
+				outputChan <- err
+				doneChan2 <- struct{}{}
+			} else {
+				errorChan1 = nil // no more selected case on this one
+				break
+			}
+		case err, errorChan2Open = <-errorChan2:
+			if errorChan2Open {
+				doneChan1 <- struct{}{}
+				outputChan <- err
+			} else {
+				errorChan2 = nil // no more selected case on this one
+				break
+			}
+		}
+	}
+}
+
+// fetchBundles waits on a channel of key batches and outputs batches of descriptors corresponding to these keys
+func fetchBundles(repo string, store storage.Store, settings CoreSettings,
+	keysChan <-chan []string,
+	batchChan chan<- model.BundleDescriptors, errorChan chan<- error,
+	doneChan <-chan struct{}, wg *sync.WaitGroup) {
+	defer func() {
+		close(batchChan)
+		close(errorChan)
+		wg.Done()
+	}()
+
+	for {
+		select {
+		case <-doneChan:
+			errorChan <- ErrInterrupted
+			return
+		case keyBatch, ok := <-keysChan:
+			if !ok {
+				return
+			}
+			batch, err := fetchBundleBatch(repo, store, settings, keyBatch)
+			if err != nil {
+				errorChan <- err
+				return
+			}
+			// send out a single batch of (ordered) bundle descriptors
+			batchChan <- batch
+		}
+	}
+}
+
+// fetchKeys fetches keys for bundles in batches, then close the keyBatchChan channel upon completion or error.
+func fetchKeys(repo string, store storage.Store, settings CoreSettings,
+	keyBatchChan chan<- []string, errorChan chan<- error,
+	doneChan <-chan struct{}, wg *sync.WaitGroup) {
+	defer func() {
+		close(keyBatchChan)
+		close(errorChan)
+		wg.Done()
+	}()
+
+	var (
+		ks   []string
+		next string
+		err  error
+	)
+
+	for {
+		// get a batch of keys
+		ks, next, err = store.KeysPrefix(context.Background(), next, model.GetArchivePathPrefixToBundles(repo), "/", settings.bundleBatchSize)
+		if err != nil {
+			select {
+			case errorChan <- err:
+				break
+			case <-doneChan:
```

Comment:
here, we don't preserve the meaning of "interrupt" in the sense of `watchForInterrupts` via, e.g.,

```
case _, interrupt <-doneChan:
	if interrupt { errorChan <- ErrInterrupted }
```

for some reason?